File: backend/src/email/mailpitClient.ts
/**
 * Minimal client for Mailpit's REST API.
 *
 * Mailpit's webhook payload is only a message *summary* (it carries the message
 * ID and a From header, but not the body or attachment bytes). So ingestion
 * calls back here to fetch the full message and download the PDF part.
 * API reference: http://<mailpit>/api/v1/
 */
import { config } from "../config.js";

export interface MailpitAddress {
  Name: string;
  Address: string;
}

export interface MailpitAttachment {
  PartID: string;
  FileName: string;
  ContentType: string;
  ContentID: string;
  Size: number;
}

/** Shape of GET /api/v1/message/{ID} (only the fields we use). */
export interface MailpitMessage {
  ID: string;
  MessageID: string;
  From: MailpitAddress | null;
  To: MailpitAddress[];
  Subject: string;
  Text: string;
  HTML: string;
  Attachments: MailpitAttachment[];
}

/** GET /api/v1/message/{ID} — full message incl. body text + attachment list. */
export async function getMessage(id: string): Promise<MailpitMessage> {
  const res = await fetch(`${config.mailpit.apiUrl}/api/v1/message/${encodeURIComponent(id)}`);
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`Mailpit GET message ${id} failed (${res.status}): ${body}`);
  }
  return (await res.json()) as MailpitMessage;
}

/** GET /api/v1/message/{ID}/part/{PartID} — raw bytes of a single attachment. */
export async function getAttachment(id: string, partId: string): Promise<Buffer> {
  const res = await fetch(
    `${config.mailpit.apiUrl}/api/v1/message/${encodeURIComponent(id)}/part/${encodeURIComponent(partId)}`,
  );
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`Mailpit GET part ${partId} failed (${res.status}): ${body}`);
  }
  return Buffer.from(await res.arrayBuffer());
}

File: backend-py/app/email/mailpit_client.py
"""Minimal client for Mailpit's REST API (port of mailpitClient.ts).

The webhook payload is only a message summary; ingestion calls back here to fetch
the full message and download the PDF part.
"""

from __future__ import annotations

from urllib.parse import quote

import httpx

from app.config import settings


async def get_message(client: httpx.AsyncClient, message_id: str) -> dict:
    """GET /api/v1/message/{ID} — full message incl. body text + attachment list."""
    resp = await client.get(
        f"{settings.mailpit_api_url}/api/v1/message/{quote(message_id, safe='')}"
    )
    if resp.status_code != 200:
        raise RuntimeError(
            f"Mailpit GET message {message_id} failed ({resp.status_code}): {resp.text}"
        )
    return resp.json()


async def get_attachment(
    client: httpx.AsyncClient, message_id: str, part_id: str
) -> bytes:
    """GET /api/v1/message/{ID}/part/{PartID} — raw bytes of a single attachment."""
    resp = await client.get(
        f"{settings.mailpit_api_url}/api/v1/message/{quote(message_id, safe='')}"
        f"/part/{quote(part_id, safe='')}"
    )
    if resp.status_code != 200:
        raise RuntimeError(
            f"Mailpit GET part {part_id} failed ({resp.status_code}): {resp.text}"
        )
    return resp.content
